### DISCUSSION TITLE
VAPIX: firmwaremanagement.cgi

### DISCUSSION THREAD

**Discussion by tonylom3** [AxisCommunications/discussions#46] [VAPIX] [answered]
Hello all - long time reader, first time poster here. I needed to get the firmware version of a subset of camera IPs. I thought it would be straight forward. But sadly, not so, at least for me. I tried using a version of @mattias-kindborg-at-work python examples he had put in a different thread. But I can't seem to get it right. Anyone have any suggestions? Doesn't have to be python.

Thanks in advance!
-Tony Lombardo

=====================================
`import requests
import sys

if len(sys.argv) != 4:
    print("Usage: list-camera-parameters.py <ip> <username> <password>")
    sys.exit(1)

ip = sys.argv[1]
username = sys.argv[2]
password = sys.argv[3]

payload = {
  "apiVersion": "1.4",
  "context": "abc",
  "method": "upgrade"
}

auth = requests.auth.HTTPDigestAuth(username, password)
r = requests.post("http://{}/axis-cgi/firmwaremanagement.cgi, data=payload, auth=auth")

print(r.status_code)
print(r.text)`

**Comment by mattias-kindborg-at-work** [comment]
Hi @tonylom3.

What language would you prefer the sample code to be written in?

**Reply by tonylom3** [reply]
Good Morning @mattias-kindborg-at-work and thank you for the reply. I am most familiar with python, but it could be bash/curl or PowerShell. Whichever is easier, I just appreciate the help! :-)

-Tony

**Reply by mattias-kindborg-at-work** [reply]
Would you please provide the command you are executing (with redacted credentials) and the error you get?

**Comment by tonylom3** [comment]
Hi Mattias, I was trying to follow the Firmware status example listed here:
https://www.axis.com/vapix-library/subjects/t10175981/section/t10118139/display?section=t10118139-t10118146

==================================

`import requests

url = "https://10.xx.xx.xx/axis-cgi/firmwaremanagement.cgi"

payload = {
    "apiVersion": "1.4",
    "context": "abc",
    "method": "status"
}
headers = {
    "Content-Type": "application/json",
    "Authorization": "Basic abcdefg9vdDp1ITlqRiYlUzEmRmllOXQx"
}

response = requests.request("POST", url, json=payload, headers=headers)

print(response.text)`

=====================================

This server could not verify that you are authorized to access the document requested. Either you supplied the wrong credentials (e.g., bad password), or your browser doesn't understand how to supply the credentials required.

**Comment by mattias-kindborg-at-work** [comment, marked as answer]
Hi @tonylom3.

I think the following code would work.

```python
import requests
import sys

if len(sys.argv) != 4:
    print("Usage: firmware-status.py <ip> <username> <password>")
    sys.exit(1)

ip = sys.argv[1]
username = sys.argv[2]
password = sys.argv[3]

auth = requests.auth.HTTPDigestAuth(username, password)
body = {
    "apiVersion": "1.4",
    "method": "status"
}
r = requests.post("http://{}/axis-cgi/firmwaremanagement.cgi".format(ip), json=body,  auth=auth)

print(r.status_code)
print(r.text)
```

When I call this I get the following response.

```sh
python3 firmware-status.py 192.168.1.80 root mypassword
200
{
  "apiVersion": "1.4",
  "method": "status",
  "data": {
    "activeFirmwareVersion": "10.9.0",
    "activeFirmwarePart": "6337444343"
  }
}
```

**Reply by tonylom3** [reply]
Success! Works beautifully, you are a scholar and a gentleman! Thank you once again for the help, I can now build on this and iterate through a bunch of IPs and get what is needed.

-Tony
